src/algokit_utils/asset.py

```diff
@@ -0,0 +1,153 @@
+import logging
+from typing import TYPE_CHECKING
+
+from algosdk.atomic_transaction_composer import AtomicTransactionComposer, TransactionWithSigner
+from algosdk.constants import TX_GROUP_LIMIT
+from algosdk.transaction import AssetTransferTxn
+
+if TYPE_CHECKING:
+    from algosdk.v2client.algod import AlgodClient
+
+from algokit_utils.models import Account
+
+__all__ = ["opt_in", "opt_out"]
+logger = logging.getLogger(__name__)
+
+
+def _ensure_asset_balance(algod_client: "AlgodClient", account: Account, asset_ids: list) -> None:
```

Comment:
Same comment as in utils-ts version.

This method and the one below are redundant, both are almost identical except for min balance parameters, you can combine them into a single method called `ensure_asset_balance_conditions` and take a third param of custom enum type call ValidationType (with option OptIn or OptOut), based on this enum you can them figure out whether to check if balance is >= 0 (opted in) or asset should not be inside accountInfo response at all (opted out).

Also why are you using account_asset_info in this methods while in utils-ts you rely on account info? Py sdk also has account info and you can potentially use that instead, its much more performant and will be consistent with your utils -ts implementation. Otherwise you are sending O(n) API calls per each asset id, while using account_info is just a single api call giving you all opted in assets on this account